src/index.js

```diff
@@ -106,40 +107,59 @@ class DropdownTreeSelect extends Component {
   }
 
   componentWillUnmount() {
-    document.removeEventListener('click', this.handleOutsideClick, false)
+    document.removeEventListener('click', this.handleDropdownCollapse, false)
   }
 
   componentWillReceiveProps(nextProps) {
     this.initNewProps(nextProps)
   }
 
+  onBlur = () => {
+    // setTimeout runs afterwards in the event.
+    // If onFocus is triggered immediately in a child component, clearTimeout will stop setTimeout to run.
+    this._timeoutID = setTimeout(() => {
```

Comment:
setTimeout/clearTimeout seems like a fragile solution. Not just from a timing perspective but I think it makes it harder to move towards hooks. I was going to ask if you have you looked at using requestAnimationFrame in lieu of setTimeout but then I see that it's being used in conjunction with clearTimeout so RAF might not be an option.

I'm curious though - why take on the additional burden of keeping track of things instead of detecting outside clicks?

Reply:
I agree it looks a bit like smelly code, but it does not delay in terms of time but on one cycle on the event loop. 

I choose this solution, first, because it builds over the existing onFocus/blur events from the dom and it doesn't need to treat corner cases. in my opinion this is easier to maintain then associating the blur focus with other events like opening and closing the dropdown. Secondly,  the show stopper for me was the complexity of having the blur triggered when using keyboard navigation in a form. you could be listening on tab on the last element but you can not be sure which one is it,you have to check the props. it seemed to add to much complexity on the code.

i also tried attaching a handler like it is done for closing the dropdown when clicking outside but i ran into some corner cases, and i am sure i cannot think of all of them. also it might be harder to maintain.